File: retrocpu_test_framework/src/tms9995/cru_handshake.ts
/**
 * TMS9995 の CRU ボード周辺モック（ハンドシェイク + 割り込み要因）。
 * 根拠: TMS9995_CPUボードメモリ_IOマップ.mdc
 * - 0x0010..0x0012: INTERRUPT_BUSY / INT1_CAUSE / INT2_CAUSE
 * - 0x0020..0x0027: ハンドシェイク制御・DATA ラッチ
 */

import type {
  Tms9995CruActor,
  Tms9995CruBit,
  Tms9995CruCpuInSignal,
  Tms9995CruCpuOutSignal,
  Tms9995CruHandshakeOptions,
  Tms9995CruHandshakeSnapshot,
  Tms9995CruReadLog,
  Tms9995CruSignalName,
  Tms9995CruWriteLog,
} from "./types.js";

const BIT_MIN = 0x0010;
const BIT_MAX = 0x0027;

const IRQ_BITS = {
  INTERRUPT_BUSY: 0x0010,
  INT1_CAUSE: 0x0011,
  INT2_CAUSE: 0x0012,
} as const;

const CPU_OUT_SIGNALS: Record<Tms9995CruCpuOutSignal, number> = {
  INTERRUPT_BUSY: IRQ_BITS.INTERRUPT_BUSY,
  HSHK_OUT_REQ: 0x0020,
  HSHK_OUT_DENA: 0x0021,
  HSHK_IN_DACK: 0x0022,
};

const CPU_IN_SIGNALS: Record<Tms9995CruCpuInSignal, number> = {
  INT1_CAUSE: IRQ_BITS.INT1_CAUSE,
  INT2_CAUSE: IRQ_BITS.INT2_CAUSE,
  HSHK_IN_REQ: 0x0024,
  HSHK_IN_DENA: 0x0025,
  HSHK_OUT_DACK: 0x0026,
};

const OUT_DATA_BASE = 0x0023;
const IN_DATA_BASE = 0x0027;

const CPU_WRITABLE = new Set<number>([
  CPU_OUT_SIGNALS.INTERRUPT_BUSY,
  CPU_OUT_SIGNALS.HSHK_OUT_REQ,
  CPU_OUT_SIGNALS.HSHK_OUT_DENA,
  CPU_OUT_SIGNALS.HSHK_IN_DACK,
  OUT_DATA_BASE,
]);

const IO_WRITABLE = new Set<number>([
  CPU_IN_SIGNALS.INT1_CAUSE,
  CPU_IN_SIGNALS.INT2_CAUSE,
  CPU_IN_SIGNALS.HSHK_IN_REQ,
  CPU_IN_SIGNALS.HSHK_IN_DENA,
  CPU_IN_SIGNALS.HSHK_OUT_DACK,
  IN_DATA_BASE,
]);

function range(from: number, to: number): number[] {
  const out: number[] = [];
  for (let p = from; p <= to; p += 1) {
    out.push(p);
  }
  return out;
}

function toBit(value: number): Tms9995CruBit {
  return value === 0 ? 0 : 1;
}

function asByte(value: number): number {
  if (!Number.isInteger(value) || value < 0 || value > 0xff) {
    throw new Error(`byte must be 0..255 (got ${value})`);
  }
  return value;
}

function fmtBitAddr(bitAddr: number): string {
  return `0x${bitAddr.toString(16).toUpperCase().padStart(4, "0")}`;
}

/**
 * CRU ハンドシェイク／割り込み要因モック。
 * CPU 側と IO 側で役割外の読み書きを検出できる。
 */
export class Tms9995CruHandshakeMock {
  readonly writes: Tms9995CruWriteLog[] = [];
  readonly reads: Tms9995CruReadLog[] = [];

  private readonly strictRoles: boolean;
  private readonly bits = new Map<number, Tms9995CruBit>();
  private outDataByte = 0;
  private inDataByte = 0;

  constructor(options: Tms9995CruHandshakeOptions = {}) {
    this.strictRoles = options.strictRoles !== false;
    for (const bitAddr of range(BIT_MIN, BIT_MAX)) {
      this.bits.set(bitAddr, 0);
    }
  }

  /** すべての線を 0 に戻し、ログも消す。 */
  reset(): void {
    for (const bitAddr of this.bits.keys()) {
      this.bits.set(bitAddr, 0);
    }
    this.outDataByte = 0;
    this.inDataByte = 0;
    this.writes.length = 0;
    this.reads.length = 0;
  }

  /**
   * 1bit 書き込み。
   * @param actor 主体（cpu / io）
   * @param bitAddr CRU ビットアドレス
   * @param value 0/1
   */
  writeBit(
    actor: Tms9995CruActor,
    bitAddr: number,
    value: Tms9995CruBit,
  ): void {
    const addr = this.normalizeBitAddr(bitAddr);
    if (this.strictRoles && !this.canWrite(actor, addr)) {
      throw new Error(
        `${actor} cannot write CRU bit ${fmtBitAddr(addr)} in handshake region`,
      );
    }
    if (addr === OUT_DATA_BASE) {
      this.outDataByte = (this.outDataByte & ~1) | (value & 1);
    } else if (addr === IN_DATA_BASE) {
      this.inDataByte = (this.inDataByte & ~1) | (value & 1);
    }
    this.bits.set(addr, value);
    this.writes.push({ actor, bitAddr: addr, value });
  }

  /**
   * 1bit 読み出し。
   * @param actor 主体（cpu / io）
   * @param bitAddr CRU ビットアドレス
   * @returns 0/1
   */
  readBit(actor: Tms9995CruActor, bitAddr: number): Tms9995CruBit {
    const addr = this.normalizeBitAddr(bitAddr);
    if (this.strictRoles && !this.canRead(actor, addr)) {
      throw new Error(
        `${actor} cannot read CRU bit ${fmtBitAddr(addr)} in handshake region`,
      );
    }
    let value: Tms9995CruBit;
    if (addr === OUT_DATA_BASE) {
      value = toBit(this.outDataByte & 1);
    } else if (addr === IN_DATA_BASE) {
      value = toBit(this.inDataByte & 1);
    } else {
      value = this.bits.get(addr) ?? 0;
    }
    this.reads.push({ actor, bitAddr: addr, value });
    return value;
  }

  /** CPU 出力信号を 1bit 書き込む。 */
  cpuWriteSignal(signal: Tms9995CruCpuOutSignal, value: Tms9995CruBit): void {
    this.writeBit("cpu", CPU_OUT_SIGNALS[signal], value);
  }

  /** CPU 入力信号を 1bit 読み出す。 */
  cpuReadSignal(signal: Tms9995CruCpuInSignal): Tms9995CruBit {
    return this.readBit("cpu", CPU_IN_SIGNALS[signal]);
  }

  /** IO 側が CPU 入力信号を 1bit 書き込む。 */
  ioWriteSignal(signal: Tms9995CruCpuInSignal, value: Tms9995CruBit): void {
    this.writeBit("io", CPU_IN_SIGNALS[signal], value);
  }

  /** IO 側が CPU 出力信号を 1bit 読み出す。 */
  ioReadSignal(signal: Tms9995CruCpuOutSignal): Tms9995CruBit {
    return this.readBit("io", CPU_OUT_SIGNALS[signal]);
  }

  /**
   * INT1 要因を IO がセットする（1=ハンドシェイク）。
   * @param cause 下位 1bit
   */
  ioSetInt1Cause(cause: number): void {
    this.writeBit("io", IRQ_BITS.INT1_CAUSE, toBit(cause & 1));
  }

  /**
   * INT1 要因を CPU 視点で読む。
   * @returns 0 または 1
   */
  cpuReadInt1Cause(): Tms9995CruBit {
    return this.readBit("cpu", IRQ_BITS.INT1_CAUSE);
  }

  /**
   * INT2 要因（0=break / 1=step）を IO がセットする。
   * @param cause 0 または 1
   */
  ioSetInt2Cause(cause: number): void {
    this.writeBit("io", IRQ_BITS.INT2_CAUSE, toBit(cause & 1));
  }

  /** INT2 要因を CPU 視点で読む。 */
  cpuReadInt2Cause(): Tms9995CruBit {
    return this.readBit("cpu", IRQ_BITS.INT2_CAUSE);
  }

  /** CPU→IO データラッチへ 1 バイトを載せる（LDCR #8 相当）。 */
  cpuWriteOutDataByte(value: number): void {
    const b = asByte(value);
    this.outDataByte = b;
    this.bits.set(OUT_DATA_BASE, toBit(b & 1));
    this.writes.push({
      actor: "cpu",
      bitAddr: OUT_DATA_BASE,
      value: toBit(b & 1),
    });
  }

  /** IO 側が CPU→IO データラッチを 1 バイトとして読む。 */
  ioReadOutDataByte(): number {
    this.reads.push({
      actor: "io",
      bitAddr: OUT_DATA_BASE,
      value: toBit(this.outDataByte & 1),
    });
    return this.outDataByte & 0xff;
  }

  /** IO→CPU データラッチへ 1 バイトを載せる。 */
  ioWriteInDataByte(value: number): void {
    const b = asByte(value);
    this.inDataByte = b;
    this.bits.set(IN_DATA_BASE, toBit(b & 1));
    this.writes.push({
      actor: "io",
      bitAddr: IN_DATA_BASE,
      value: toBit(b & 1),
    });
  }

  /** CPU 側が IO→CPU データラッチを 1 バイトとして読む（STCR #8 相当）。 */
  cpuReadInDataByte(): number {
    this.reads.push({
      actor: "cpu",
      bitAddr: IN_DATA_BASE,
      value: toBit(this.inDataByte & 1),
    });
    return this.inDataByte & 0xff;
  }

  /** 任意の信号アドレスを名前で引く（テスト用）。 */
  bitAddrOf(signal: Tms9995CruSignalName): number {
    if (signal in CPU_OUT_SIGNALS) {
      const key = signal as Tms9995CruCpuOutSignal;
      return CPU_OUT_SIGNALS[key];
    }
    const key = signal as Tms9995CruCpuInSignal;
    return CPU_IN_SIGNALS[key];
  }

  /** 現在の線状態を取り出す。 */
  snapshot(): Tms9995CruHandshakeSnapshot {
    const cpuOutSignals = {
      INTERRUPT_BUSY: this.peekBit(CPU_OUT_SIGNALS.INTERRUPT_BUSY),
      HSHK_OUT_REQ: this.peekBit(CPU_OUT_SIGNALS.HSHK_OUT_REQ),
      HSHK_OUT_DENA: this.peekBit(CPU_OUT_SIGNALS.HSHK_OUT_DENA),
      HSHK_IN_DACK: this.peekBit(CPU_OUT_SIGNALS.HSHK_IN_DACK),
    };
    const cpuInSignals = {
      INT1_CAUSE: this.peekBit(CPU_IN_SIGNALS.INT1_CAUSE),
      INT2_CAUSE: this.peekBit(CPU_IN_SIGNALS.INT2_CAUSE),
      HSHK_IN_REQ: this.peekBit(CPU_IN_SIGNALS.HSHK_IN_REQ),
      HSHK_IN_DENA: this.peekBit(CPU_IN_SIGNALS.HSHK_IN_DENA),
      HSHK_OUT_DACK: this.peekBit(CPU_IN_SIGNALS.HSHK_OUT_DACK),
    };
    const bits: Record<string, Tms9995CruBit> = {};
    for (const bitAddr of range(BIT_MIN, BIT_MAX)) {
      if (bitAddr === OUT_DATA_BASE) {
        bits[fmtBitAddr(bitAddr)] = toBit(this.outDataByte & 1);
      } else if (bitAddr === IN_DATA_BASE) {
        bits[fmtBitAddr(bitAddr)] = toBit(this.inDataByte & 1);
      } else {
        bits[fmtBitAddr(bitAddr)] = this.peekBit(bitAddr);
      }
    }
    return {
      cpuOutSignals,
      cpuInSignals,
      outDataByte: this.outDataByte & 0xff,
      inDataByte: this.inDataByte & 0xff,
      bits,
    };
  }

  private peekBit(bitAddr: number): Tms9995CruBit {
    return this.bits.get(bitAddr) ?? 0;
  }

  private normalizeBitAddr(bitAddr: number): number {
    if (!Number.isInteger(bitAddr)) {
      throw new Error(`bit address must be integer (got ${String(bitAddr)})`);
    }
    if (bitAddr < BIT_MIN || bitAddr > BIT_MAX) {
      throw new Error(
        `bit address ${fmtBitAddr(bitAddr)} is out of handshake region (${fmtBitAddr(
          BIT_MIN,
        )}..${fmtBitAddr(BIT_MAX)})`,
      );
    }
    return bitAddr;
  }

  private canWrite(actor: Tms9995CruActor, bitAddr: number): boolean {
    return actor === "cpu"
      ? CPU_WRITABLE.has(bitAddr)
      : IO_WRITABLE.has(bitAddr);
  }

  private canRead(actor: Tms9995CruActor, bitAddr: number): boolean {
    return actor === "cpu"
      ? IO_WRITABLE.has(bitAddr)
      : CPU_WRITABLE.has(bitAddr);
  }
}

/** CRU 領域の範囲（両端含む）。ハンドシェイク + 割り込み要因。 */
export const TMS9995_CRU_HANDSHAKE_REGION = {
  bitAddrMin: BIT_MIN,
  bitAddrMax: BIT_MAX,
  outDataBase: OUT_DATA_BASE,
  inDataBase: IN_DATA_BASE,
  irqBusy: IRQ_BITS.INTERRUPT_BUSY,
  int1Cause: IRQ_BITS.INT1_CAUSE,
  int2Cause: IRQ_BITS.INT2_CAUSE,
} as const;

/** 役割名と CRU ビットアドレスの対応。 */
export const TMS9995_CRU_HANDSHAKE_SIGNALS = {
  ...CPU_OUT_SIGNALS,
  ...CPU_IN_SIGNALS,
  HSHK_OUT_DATA: OUT_DATA_BASE,
  HSHK_IN_DATA: IN_DATA_BASE,
} as const;

File: retrocpu_test_framework_rs/src/framework/tms9995/cru_handshake.rs
//! TMS9995 CRU ハンドシェイク／割り込み要因モック。
//!
//! 根拠: `retrocpu_test_framework_ts/src/tms9995/cru_handshake.ts`

use std::collections::HashMap;

use crate::error::FrameworkError;

const BIT_MIN: u16 = 0x0010;
const BIT_MAX: u16 = 0x0027;

const IRQ_INTERRUPT_BUSY: u16 = 0x0010;
const IRQ_INT1_CAUSE: u16 = 0x0011;
const IRQ_INT2_CAUSE: u16 = 0x0012;

const CPU_OUT_HSHK_OUT_REQ: u16 = 0x0020;
const CPU_OUT_HSHK_OUT_DENA: u16 = 0x0021;
const CPU_OUT_HSHK_IN_DACK: u16 = 0x0022;
const CPU_OUT_DATA: u16 = 0x0023;

const CPU_IN_HSHK_IN_REQ: u16 = 0x0024;
const CPU_IN_HSHK_IN_DENA: u16 = 0x0025;
const CPU_IN_HSHK_OUT_DACK: u16 = 0x0026;
const CPU_IN_DATA: u16 = 0x0027;

/// CRU ハンドシェイク領域。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Tms9995CruHandshakeRegion {
    pub bit_addr_min: u16,
    pub bit_addr_max: u16,
}

pub const TMS9995_CRU_HANDSHAKE_REGION: Tms9995CruHandshakeRegion = Tms9995CruHandshakeRegion {
    bit_addr_min: BIT_MIN,
    bit_addr_max: BIT_MAX,
};

/// 信号名 → CRU ビットアドレス。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Tms9995CruHandshakeSignals {
    pub interrupt_busy: u16,
    pub int1_cause: u16,
    pub int2_cause: u16,
    pub hshk_out_req: u16,
    pub hshk_out_dena: u16,
    pub hshk_in_dack: u16,
    pub hshk_in_req: u16,
    pub hshk_in_dena: u16,
    pub hshk_out_dack: u16,
}

pub const TMS9995_CRU_HANDSHAKE_SIGNALS: Tms9995CruHandshakeSignals = Tms9995CruHandshakeSignals {
    interrupt_busy: IRQ_INTERRUPT_BUSY,
    int1_cause: IRQ_INT1_CAUSE,
    int2_cause: IRQ_INT2_CAUSE,
    hshk_out_req: CPU_OUT_HSHK_OUT_REQ,
    hshk_out_dena: CPU_OUT_HSHK_OUT_DENA,
    hshk_in_dack: CPU_OUT_HSHK_IN_DACK,
    hshk_in_req: CPU_IN_HSHK_IN_REQ,
    hshk_in_dena: CPU_IN_HSHK_IN_DENA,
    hshk_out_dack: CPU_IN_HSHK_OUT_DACK,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Tms9995CruActor {
    Cpu,
    Io,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tms9995CruHandshakeOptions {
    pub strict_roles: bool,
}

impl Default for Tms9995CruHandshakeOptions {
    fn default() -> Self {
        Self {
            strict_roles: true,
        }
    }
}

fn fmt_bit_addr(bit_addr: u16) -> String {
    format!("0x{bit_addr:04X}")
}

fn cpu_writable(bit: u16) -> bool {
    matches!(
        bit,
        IRQ_INTERRUPT_BUSY | CPU_OUT_HSHK_OUT_REQ | CPU_OUT_HSHK_OUT_DENA | CPU_OUT_HSHK_IN_DACK | CPU_OUT_DATA
    )
}

fn io_writable(bit: u16) -> bool {
    matches!(
        bit,
        IRQ_INT1_CAUSE
            | IRQ_INT2_CAUSE
            | CPU_IN_HSHK_IN_REQ
            | CPU_IN_HSHK_IN_DENA
            | CPU_IN_HSHK_OUT_DACK
            | CPU_IN_DATA
    )
}

fn cpu_readable(bit: u16) -> bool {
    io_writable(bit)
}

fn io_readable(bit: u16) -> bool {
    cpu_writable(bit)
}

/// CRU ハンドシェイクモック。
#[derive(Debug, Clone)]
pub struct Tms9995CruHandshakeMock {
    strict_roles: bool,
    bits: HashMap<u16, u8>,
    out_data_byte: u8,
    in_data_byte: u8,
}

impl Default for Tms9995CruHandshakeMock {
    fn default() -> Self {
        Self::new(Tms9995CruHandshakeOptions::default())
    }
}

impl Tms9995CruHandshakeMock {
    /// オプション付きで作る。
    pub fn new(options: Tms9995CruHandshakeOptions) -> Self {
        Self {
            strict_roles: options.strict_roles,
            bits: HashMap::new(),
            out_data_byte: 0,
            in_data_byte: 0,
        }
    }

    fn check_range(bit: u16) -> Result<(), FrameworkError> {
        if !(BIT_MIN..=BIT_MAX).contains(&bit) {
            return Err(FrameworkError::invalid_argument(format!(
                "bit address out of handshake region: {}",
                fmt_bit_addr(bit)
            )));
        }
        Ok(())
    }

    /// 任意ビットを読む。
    pub fn read_bit(&self, actor: Tms9995CruActor, bit: u16) -> Result<u8, FrameworkError> {
        Self::check_range(bit)?;
        if self.strict_roles {
            let ok = match actor {
                Tms9995CruActor::Cpu => cpu_readable(bit),
                Tms9995CruActor::Io => io_readable(bit),
            };
            if !ok {
                return Err(FrameworkError::invalid_argument(format!(
                    "{:?} cannot read {}",
                    actor,
                    fmt_bit_addr(bit)
                )));
            }
        }
        if bit == CPU_OUT_DATA {
            return Ok(self.out_data_byte & 1);
        }
        if bit == CPU_IN_DATA {
            return Ok(self.in_data_byte & 1);
        }
        Ok(*self.bits.get(&bit).unwrap_or(&0))
    }

    /// 任意ビットを書く。
    pub fn write_bit(&mut self, actor: Tms9995CruActor, bit: u16, value: u8) -> Result<(), FrameworkError> {
        Self::check_range(bit)?;
        if value != 0 && value != 1 {
            return Err(FrameworkError::invalid_argument(format!(
                "bit must be 0 or 1 (got {value})"
            )));
        }
        if self.strict_roles {
            let ok = match actor {
                Tms9995CruActor::Cpu => cpu_writable(bit),
                Tms9995CruActor::Io => io_writable(bit),
            };
            if !ok {
                return Err(FrameworkError::invalid_argument(format!(
                    "{:?} cannot write {}",
                    actor,
                    fmt_bit_addr(bit)
                )));
            }
        }
        if bit == CPU_OUT_DATA {
            self.out_data_byte = (self.out_data_byte & !1) | (value & 1);
        } else if bit == CPU_IN_DATA {
            self.in_data_byte = (self.in_data_byte & !1) | (value & 1);
        }
        self.bits.insert(bit, value);
        Ok(())
    }

    pub fn cpu_write_signal(&mut self, bit: u16, value: u8) -> Result<(), FrameworkError> {
        self.write_bit(Tms9995CruActor::Cpu, bit, value)
    }

    pub fn io_read_signal(&self, bit: u16) -> Result<u8, FrameworkError> {
        self.read_bit(Tms9995CruActor::Io, bit)
    }

    pub fn io_write_signal(&mut self, bit: u16, value: u8) -> Result<(), FrameworkError> {
        self.write_bit(Tms9995CruActor::Io, bit, value)
    }

    pub fn cpu_read_signal(&self, bit: u16) -> Result<u8, FrameworkError> {
        self.read_bit(Tms9995CruActor::Cpu, bit)
    }

    pub fn cpu_write_out_data_byte(&mut self, value: u8) -> Result<(), FrameworkError> {
        self.out_data_byte = value;
        self.bits.insert(CPU_OUT_DATA, value & 1);
        Ok(())
    }

    pub fn io_read_out_data_byte(&self) -> u8 {
        self.out_data_byte
    }

    pub fn io_write_in_data_byte(&mut self, value: u8) -> Result<(), FrameworkError> {
        self.in_data_byte = value;
        self.bits.insert(CPU_IN_DATA, value & 1);
        Ok(())
    }

    pub fn cpu_read_in_data_byte(&self) -> u8 {
        self.in_data_byte
    }

    pub fn io_set_int1_cause(&mut self, value: u8) -> Result<(), FrameworkError> {
        self.write_bit(Tms9995CruActor::Io, IRQ_INT1_CAUSE, value & 1)
    }

    pub fn cpu_read_int1_cause(&self) -> Result<u8, FrameworkError> {
        self.cpu_read_signal(IRQ_INT1_CAUSE)
    }

    pub fn io_set_int2_cause(&mut self, value: u8) -> Result<(), FrameworkError> {
        self.write_bit(Tms9995CruActor::Io, IRQ_INT2_CAUSE, value & 1)
    }

    pub fn cpu_read_int2_cause(&self) -> Result<u8, FrameworkError> {
        self.cpu_read_signal(IRQ_INT2_CAUSE)
    }

    /// ビット線とデータバイトを初期化する。
    pub fn reset(&mut self) {
        self.bits.clear();
        self.out_data_byte = 0;
        self.in_data_byte = 0;
    }

    /// 現在状態のスナップショット。
    pub fn snapshot(&self) -> Tms9995CruHandshakeSnapshot {
        let mut bits = HashMap::new();
        for bit in BIT_MIN..=BIT_MAX {
            bits.insert(fmt_bit_addr(bit), *self.bits.get(&bit).unwrap_or(&0));
        }
        Tms9995CruHandshakeSnapshot {
            cpu_out_signals: Tms9995CruCpuOutSignals {
                hshk_out_req: *self.bits.get(&CPU_OUT_HSHK_OUT_REQ).unwrap_or(&0),
                hshk_out_dena: *self.bits.get(&CPU_OUT_HSHK_OUT_DENA).unwrap_or(&0),
                hshk_in_dack: *self.bits.get(&CPU_OUT_HSHK_IN_DACK).unwrap_or(&0),
                interrupt_busy: *self.bits.get(&IRQ_INTERRUPT_BUSY).unwrap_or(&0),
            },
            cpu_in_signals: Tms9995CruCpuInSignals {
                hshk_in_req: *self.bits.get(&CPU_IN_HSHK_IN_REQ).unwrap_or(&0),
                hshk_in_dena: *self.bits.get(&CPU_IN_HSHK_IN_DENA).unwrap_or(&0),
                hshk_out_dack: *self.bits.get(&CPU_IN_HSHK_OUT_DACK).unwrap_or(&0),
                int1_cause: *self.bits.get(&IRQ_INT1_CAUSE).unwrap_or(&0),
                int2_cause: *self.bits.get(&IRQ_INT2_CAUSE).unwrap_or(&0),
            },
            out_data_byte: self.out_data_byte,
            in_data_byte: self.in_data_byte,
            bits,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tms9995CruCpuOutSignals {
    pub hshk_out_req: u8,
    pub hshk_out_dena: u8,
    pub hshk_in_dack: u8,
    pub interrupt_busy: u8,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tms9995CruCpuInSignals {
    pub hshk_in_req: u8,
    pub hshk_in_dena: u8,
    pub hshk_out_dack: u8,
    pub int1_cause: u8,
    pub int2_cause: u8,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tms9995CruHandshakeSnapshot {
    pub cpu_out_signals: Tms9995CruCpuOutSignals,
    pub cpu_in_signals: Tms9995CruCpuInSignals,
    pub out_data_byte: u8,
    pub in_data_byte: u8,
    pub bits: HashMap<String, u8>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cpu_to_io_signal_path() {
        let mut cru = Tms9995CruHandshakeMock::default();
        cru.cpu_write_signal(CPU_OUT_HSHK_OUT_REQ, 1).expect("write");
        cru.cpu_write_signal(CPU_OUT_HSHK_OUT_DENA, 1).expect("write");
        cru.cpu_write_out_data_byte(0xa5).expect("data");
        assert_eq!(cru.io_read_signal(CPU_OUT_HSHK_OUT_REQ).expect("read"), 1);
        assert_eq!(cru.io_read_signal(CPU_OUT_HSHK_OUT_DENA).expect("read"), 1);
        assert_eq!(cru.io_read_out_data_byte(), 0xa5);
    }

    #[test]
    fn strict_roles_rejects_wrong_actor() {
        let mut cru = Tms9995CruHandshakeMock::default();
        assert!(cru
            .write_bit(Tms9995CruActor::Cpu, CPU_IN_HSHK_IN_REQ, 1)
            .is_err());
        assert!(cru
            .write_bit(Tms9995CruActor::Io, CPU_OUT_HSHK_OUT_REQ, 1)
            .is_err());
    }
}
